lib/llm_composer.ex

```diff
@@ -1,3 +1,129 @@
+    auto_exec_functions: false,
+    functions: []
+  }
+
+  # Initiate a simple chat interaction with the defined settings
+  {:ok, response} = LlmComposer.simple_chat(settings, "Hello, how are you?")
+
+  # Print the main response from the assistant
+  IO.inspect(response.main_response)
+  ```
+
+  ### Output Example
+
+  Running this code might produce the following log and output:
+
+  ```
+  16:41:07.594 [debug] input_tokens=18, output_tokens=9
+  %LlmComposer.Message{
+    type: :assistant,
+    content: "Hello! How can I assist you today?"
+  }
+  ```
+
+  In this example, the simple_chat/2 function sends the user's message to the language model using the provided settings, and the response is displayed as the assistant's reply.
+  """
+
+  alias LlmComposer.Helpers
+  alias LlmComposer.LlmResponse
+  alias LlmComposer.Message
+  alias LlmComposer.Settings
+
+  require Logger
+
+  @type messages :: [Message.t()]
+
+  @doc """
+  Initiates a simple chat interaction with the language model.
+
+  ## Parameters
+    - `settings`: The settings for the language model, including prompts and options.
+    - `msg`: The user message to be sent to the language model.
+
+  ## Returns
+    - The result of the language model's response, which may include function executions if specified.
+  """
+  @spec simple_chat(Settings.t(), String.t()) :: Helpers.action_result()
+  def simple_chat(%Settings{} = settings, msg) do
+    messages = get_messages(settings, msg, [], %{})
+
+    run_completion(settings, messages)
+  end
+
+  @doc """
+  Runs the completion process by sending messages to the language model and handling the response.
+
+  ## Parameters
+    - `settings`: The settings for the language model, including prompts, model options, and functions.
+    - `messages`: The list of messages to be sent to the language model.
+    - `previous_response` (optional): The previous response object, if any, used for context.
+
+  ## Returns
+    - A tuple containing `:ok` with the response or `:error` if the model call fails.
+  """
+  @spec run_completion(Settings.t(), messages(), LlmResponse.t() | nil) :: Helpers.action_result()
+  def run_completion(settings, messages, previous_response \\ nil) do
+    system_msg = Message.new(:system, settings.system_prompt)
+    model_opts = Keyword.merge(settings.model_opts, functions: settings.functions)
+
+    messages
+    |> settings.model.run(system_msg, model_opts)
+    |> then(fn
+      {:ok, res} ->
+        # set previous response all the time
+        res = %LlmResponse{res | previous_response: previous_response}
+
+        Logger.debug("input_tokens=#{res.input_tokens}, output_tokens=#{res.output_tokens}")
+
+        if settings.auto_exec_functions do
+          maybe_run_functions(res, messages, settings)
+        else
+          {:ok, res}
+        end
+
+      {:error, data} = resp ->
+        Logger.error("error in llm call: #{inspect(data)}")
+        resp
+    end)
+  end
+
+  @doc false
```

Comment:
Do we need to add `@doc false` in private functions?

Reply:
I think not, gpt-4o did that for me, gonna remove it